pkg/metrics/metrics.go

```diff
@@ -16,45 +16,98 @@ package metrics
 
 import (
 	"context"
-	"net/http"
-	"strconv"
+	"fmt"
 
 	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
 
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 
 var log = logf.Log.WithName("metrics")
 
-// ExposeMetricsPort generate a Kubernetes Service to expose metrics port
-func ExposeMetricsPort() *v1.Service {
-	http.Handle("/"+k8sutil.PrometheusMetricsPortName, promhttp.Handler())
-	go http.ListenAndServe(":"+strconv.Itoa(k8sutil.PrometheusMetricsPort), nil)
+// PrometheusPortName defines the port name used in kubernetes deployment and service resources
+const PrometheusPortName = "metrics"
 
-	service, err := k8sutil.InitOperatorService()
+// ExposeMetricsPort creates a Kubernetes Service to expose the passed metrics port.
+func ExposeMetricsPort(port int32, mgr manager.Manager) (*v1.Service, error) {
+	// We do not need to check the validity of the port, as controller-runtime
+	// would error out and we would never get to this stage.
+	s, err := initOperatorService(port, PrometheusPortName)
 	if err != nil {
-		log.Error(err, "Failed to initialize service object for operator metrics")
-		return nil
+		return nil, fmt.Errorf("failed to initialize service object for metrics: %v", err)
```

Comment:
`initOperatorService()` can return [`ErrNoNamespace`](https://github.com/operator-framework/operator-sdk/blob/f2f2e916cc7a35516b77abcc0e8a2e938ed0bc91/pkg/k8sutil/k8sutil.go#L40) via `GetOperatorNamespace()` when the operator is running locally via `up local`.
Rightfully so since we don't want to create the metrics service when running locally. Esp once we add the ownerref to the service.

Since the operator running locally is a common case I think `ExposeMetricsPort()` should not return an error and instead log that it's skipping the service setup since it's running locally and return nil.
Just like we do for [leader.Become()](https://github.com/operator-framework/operator-sdk/blob/master/pkg/leader/leader.go#L49-L59).

I can see we already ignore all errors for `ExposeMetricsPort()` by just logging it in `cmd/main.go` but I think it's worth differentiating this case from any other errors.

Reply:
Yes that is a better idea! sgtm 👍